controller/db/controller_setting_store.go

```diff
@@ -0,0 +1,375 @@
+/*
+	Copyright NetFoundry Inc.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package db
+
+import (
+	"context"
+	"errors"
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/foundation/v2/errorz"
+	"github.com/openziti/foundation/v2/stringz"
+	"github.com/openziti/storage/boltz"
+	"go.etcd.io/bbolt"
+	"reflect"
+)
+
+const (
+	ControllerSettingGlobalId = "global"
+	ControllerSettingAny      = "any"
+
+	FieldControllerSettingOidc = "oidc"
+
+	FieldControllerSettingOidcRedirectUris   = FieldControllerSettingOidc + ".redirectUris"
+	FieldControllerSettingOidcPostLogoutUris = FieldControllerSettingOidc + ".postLogoutUris"
+	FieldControllerSettingIsMigration        = "isMigration"
+)
+
+type controllerFieldChangeContextKey string
+
+const controllerFieldChangeContext controllerFieldChangeContextKey = "controllerFieldChangeContext"
+
+type ControllerSettingDef interface {
+	Id() string
+	FillEntity(bucket *boltz.TypedBucket)
+	PersistEntity(ctx *boltz.PersistContext, bucket *boltz.TypedBucket, entity *ControllerSetting)
+}
+
+type ControllerSetting struct {
+	boltz.BaseExtEntity
+	Oidc *OidcSettingDef
+
+	IsMigration bool
+}
+
+func (entity *ControllerSetting) GetEntityType() string {
+	return EntityTypeSettings
+}
+
+// MergeSettings returns a new setting instance with the global settings overridden by the provided settings
+func (entity *ControllerSetting) MergeSettings(setting *ControllerSetting) *ControllerSetting {
+	newCs := &ControllerSetting{
+		BaseExtEntity: setting.BaseExtEntity,
+		Oidc:          entity.Oidc,
+	}
+
+	if setting.Oidc != nil {
+		newCs.Oidc = setting.Oidc
+	}
+
+	return newCs
+}
+
+var _ ControllerSettingStore = (*controllerSettingStoreImpl)(nil)
+
+type ControllerSettingStore interface {
+	Store[*ControllerSetting]
+	Watch(setting string, listener ControllerSettingListener, controllerId string, additionalControllerIds ...string)
+	CreateGlobalDefault(ctx boltz.MutateContext, settings *ControllerSetting) error
+}
+
+func newControllerSettingStore(provider DbProviderF, stores *stores) *controllerSettingStoreImpl {
+	store := &controllerSettingStoreImpl{
+		dbProvider: provider,
+	}
+	store.baseStore = newBaseStore[*ControllerSetting](stores, store)
+	store.InitImpl(store)
+	return store
+}
+
+type controllerSettingStoreImpl struct {
+	*baseStore[*ControllerSetting]
+	indexName                  boltz.ReadIndex
+	symbolEnrollments          boltz.EntitySetSymbol
```

Comment:
The map is likely being accessed/mutated across goroutines, should probably be protected